File: clients/ios/AquaSight/GuestStore.swift
import Foundation

struct GuestStore {
    private let key = "aquasight.guest"
    private let defaults = UserDefaults.standard

    private func load() -> [String: Any] {
        defaults.dictionary(forKey: key) ?? ["items": [:], "deleted": [:], "reads": [:], "prefs": [:]]
    }

    private func save(_ data: [String: Any]) {
        defaults.set(data, forKey: key)
    }

    func items() -> [[String: Any]] {
        let map = load()["items"] as? [String: [String: Any]] ?? [:]
        return Array(map.values)
    }

    func item(id: String) -> [String: Any]? {
        (load()["items"] as? [String: [String: Any]])?[id]
    }

    func deletedIds() -> [String] {
        Array((load()["deleted"] as? [String: Any] ?? [:]).keys)
    }

    func isSaved(_ id: String) -> Bool {
        item(id: id) != nil
    }

    func save(id: String, snapshot: [String: Any]) {
        var data = load()
        var items = data["items"] as? [String: [String: Any]] ?? [:]
        var deleted = data["deleted"] as? [String: Any] ?? [:]
        items[id] = snapshot
        deleted.removeValue(forKey: id)
        data["items"] = items
        data["deleted"] = deleted
        save(data)
    }

    func remove(id: String) {
        var data = load()
        var items = data["items"] as? [String: [String: Any]] ?? [:]
        var deleted = data["deleted"] as? [String: Any] ?? [:]
        items.removeValue(forKey: id)
        deleted[id] = true
        data["items"] = items
        data["deleted"] = deleted
        save(data)
    }

    func markRead(_ id: String) {
        var data = load()
        var reads = data["reads"] as? [String: String] ?? [:]
        if reads[id] == nil {
            reads[id] = ISO8601DateFormatter().string(from: Date())
            data["reads"] = reads
            save(data)
        }
    }

    func mergeBody() -> [String: Any] {
        var favorites: [[String: Any]] = []
        var seen = Set<String>()
        for id in deletedIds() {
            seen.insert(id)
            favorites.append(["id": id, "deleted": true])
        }
        for it in items() {
            guard let id = it["id"] as? String, seen.insert(id).inserted else { continue }
            favorites.append(["id": id, "snapshot": it])
        }
        let data = load()
        return [
            "reads": data["reads"] ?? [:],
            "prefs": data["prefs"] ?? [:],
            "favorites": favorites
        ]
    }
}

File: clients/android/app/src/main/java/com/aquasight/app/GuestStore.kt
package com.aquasight.app

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

class GuestStore(context: Context) {
    private val prefs = context.getSharedPreferences("aquasight.guest", Context.MODE_PRIVATE)

    fun items(): Map<String, JSONObject> {
        val out = mutableMapOf<String, JSONObject>()
        val raw = JSONObject(prefs.getString("data", "{}") ?: "{}")
        val items = raw.optJSONObject("items") ?: JSONObject()
        items.keys().forEach { id -> out[id] = items.getJSONObject(id) }
        return out
    }

    fun deletedIds(): List<String> {
        val raw = JSONObject(prefs.getString("data", "{}") ?: "{}")
        val deleted = raw.optJSONObject("deleted") ?: JSONObject()
        return deleted.keys().asSequence().toList()
    }

    fun reads(): JSONObject {
        val raw = JSONObject(prefs.getString("data", "{}") ?: "{}")
        return raw.optJSONObject("reads") ?: JSONObject()
    }

    fun prefs(): JSONObject {
        val raw = JSONObject(prefs.getString("data", "{}") ?: "{}")
        return raw.optJSONObject("prefs") ?: JSONObject()
    }

    fun save(id: String, snapshot: JSONObject) {
        mutate { data ->
            data.optJSONObject("items")?.put(id, snapshot) ?: data.put("items", JSONObject().put(id, snapshot))
            data.optJSONObject("deleted")?.remove(id)
        }
    }

    fun remove(id: String) {
        mutate { data ->
            data.optJSONObject("items")?.remove(id)
            val deleted = data.optJSONObject("deleted") ?: JSONObject().also { data.put("deleted", it) }
            deleted.put(id, true)
        }
    }

    fun markRead(id: String) {
        mutate { data ->
            val reads = data.optJSONObject("reads") ?: JSONObject().also { data.put("reads", it) }
            if (!reads.has(id)) reads.put(id, java.time.Instant.now().toString())
        }
    }

    fun mergeBody(): JSONObject {
        val favorites = JSONArray()
        for (id in deletedIds()) {
            favorites.put(JSONObject().put("id", id).put("deleted", true))
        }
        val seen = deletedIds().toMutableSet()
        for ((id, snap) in items()) {
            if (seen.add(id)) favorites.put(JSONObject().put("id", id).put("snapshot", snap))
        }
        return JSONObject()
            .put("reads", reads())
            .put("prefs", prefs())
            .put("favorites", favorites)
    }

    private fun mutate(block: (JSONObject) -> Unit) {
        val data = JSONObject(prefs.getString("data", "{}") ?: "{}")
        if (!data.has("items")) data.put("items", JSONObject())
        if (!data.has("deleted")) data.put("deleted", JSONObject())
        if (!data.has("reads")) data.put("reads", JSONObject())
        if (!data.has("prefs")) data.put("prefs", JSONObject())
        block(data)
        prefs.edit().putString("data", data.toString()).apply()
    }
}
